type_decoders/README.md

```diff
@@ -0,0 +1,787 @@
+
+### isChainOf()
+
+```ts
+interface IChainOfDecoderOptions {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isChainOf<T extends [unknown, ...unknown[]], R = ChainOfDecoderReturn<T>, I = DecoderInputType<T[0]>>(
+  decoders: { [I in keyof T]: Decoder<T[I], DecoderInputType<T[SubtractOne<I>]>> },
+  options?: IChainOfDecoderOptions
+): Decoder<R, I>;
+
+function isChainOf<T extends [unknown, ...unknown[]], R = ChainOfDecoderReturn<T>, I = DecoderInputType<T[0]>>(
+  decoders: { [I in keyof T]: Decoder<T[I], DecoderInputType<T[SubtractOne<I>]>> | PromiseDecoder<T[I], DecoderInputType<T[SubtractOne<I>]>> },
+  options?: IChainOfDecoderOptions
+): PromiseDecoder<R, I>;
+```
+
+`isChainOf()` accepts an array of decoders and attempts to decode a provided value using all of them, in order. The successful output of one decoder is provided as input to the next decoder. `isChainOf()` returns the `DecoderSuccess` value of the last decoder in the chain or `DecoderError` on the first failure.
+
+### isObject()
+
+```ts
+interface IObjectDecoderOptions<T> {
+  msg?: DecoderErrorMsgArg;
+  keyMap?: { [P in keyof T]?: string | number };
+}
+
+function isObject<T>(decoderObject: { [P in keyof T]: Decoder<T[P]> }, options?: IObjectDecoderOptions<T>): Decoder<T> ;
+function isObject<T>(decoderObject: { [P in keyof T]: Decoder<T[P]> | PromiseDecoder<T[P]> }, options?: IObjectDecoderOptions<T>): PromiseDecoder<T> ;
+```
+
+`isObject()` accepts a `key: Decoder` argument object and returns a new decoder that will verify that an input is a non-null object, that the input has the keys specified in the argument object, and that the key values pass the relevant decoder provided by the argument object. On `DecoderSuccess`, the a new object is returned and the key-values of that object are provided by output from the argument decoders.
+
+### isExactObject()
+
+```ts
+interface IExactObjectDecoderOptions<T> {
+  msg?: DecoderErrorMsgArg;
+  keyMap?: { [P in keyof T]?: string | number };
+}
+
+function isExactObject<T>(decoderObject: { [P in keyof T]: Decoder<T[P]> }, options?: IExactObjectDecoderOptions<T>): Decoder<T> ;
+function isExactObject<T>(decoderObject: { [P in keyof T]: Decoder<T[P]> | PromiseDecoder<T[P]> }, options?: IExactObjectDecoderOptions<T>): PromiseDecoder<T> ;
+```
+
+`isExactObject()` is the same as `isObject()`, except the input object cannot have any excess properties.
+
+### isDictionary()
+
+```ts
+interface IDictionaryDecoderOptions<V> {
+  msg?: DecoderErrorMsgArg;
+}
+
+function isDictionary<R, V = unknown>(decoder: Decoder<R, V>, options?: IDictionaryDecoderOptions<V>): Decoder<R[], V>;
+function isDictionary<R, V = unknown>(decoder: PromiseDecoder<R, V>, options?: IDictionaryDecoderOptions<V>): PromiseDecoder<R[], V>;
+```
+
+`isDictionary()` receives a decoder argument and uses that decoder to process all values (regardless of key) of an input object.
+
+### isArray()
```

Comment:
`isArrayOf`?

Reply:
`isArray()` actually supports being called without arguments (to test if something is an array but ignore what is inside). More broadly though, I'm unconvinced that `isObject`, `isArray`, `isTuple`, `isDictionary`, etc need an "Of" suffix for clarity?

The only reason why I added "Of" to `isAnyOf([])` and `isChainOf([])` was because they felt a little confusing without it, but that doesn't seem to be the same for `isObject` etc. For one, `isArray`, `isObject`, etc are all concrete *things* whereas `isAnyOf()` / `isChainOf()` are more abstract concepts. Definitely open to feedback here though.

Reply:
IMO `isArray` is more abstract, whereas `isTuple` is the more concrete version.